File: hosts/android/app/src/main/java/com/dshmobile/spike/SessionLiveProbe.kt
package com.dshmobile.spike

import android.webkit.WebView
import org.json.JSONArray
import org.json.JSONObject

/**
 * The b-android session-live same-origin probe (platform-side; Kotlin
 * sibling of hosts/ios SessionLiveProbe.swift): the page is untouched
 * upstream code — the probe defines `__dshSessionRun`, posts the REAL
 * session.list RPC through the official envelope, attaches the mux
 * session/journal stream for the session the runtime reports, collects the
 * journal frames (baseline + the live turn-2 stream), and reads the TRUE
 * rendered state. The Android WebView cannot await a Promise from
 * evaluateJavascript, so the run posts its JSON result through the
 * `dshProbe` JavascriptInterface (event-driven, no polling).
 */
object SessionLiveProbe {

    /** Expected journal shape (frozen from the observed on-device runs, the
     * iOS sibling included): the FIRST scripted turn's full log rides the
     * baseline — header + context folds included — and the second turn
     * streams LIVE after the attach with the shorter log. 19 frames,
     * contiguous seqs 0..18. */
    val baselineTypes = listOf(
        "agent/inbox/spliced", "turn/start", "agent/inbox/spliced", "step/start",
        "system/message", "user/message", "request/header", "request/context",
        "assistant/message", "step/end", "turn/end",
    )
    val liveTypes = listOf(
        "agent/inbox/spliced", "turn/start", "agent/inbox/spliced", "step/start",
        "user/message", "assistant/message", "step/end", "turn/end",
    )

    private const val STREAM_ID = "bandroid-probe-journal"
    private const val SHELL_MARK = "\u9009\u62e9\u5de5\u4f5c\u533a"

    /** Defines `__dshSessionRun` on the page. Step order: the official shell
     * settle, the real RPC (session.list), mux upgrade + journal attach for
     * the reported session, frame collection until two assistant/message
     * events quiet down, then the rendered read. The result JSON lands via
     * dshProbe.post. Each leg is a separate constant (each stays under the
     * function-size gate; interpolation assembles the one script). */
    fun probeScript(): String = """
        window.__dshSessionRun = async () => {
          const out = {};
          $postHelper
          $shellSettleLeg
          $sessionListLeg
          $journalLeg
          $renderedLeg
          dshProbe.post(JSON.stringify(out));
        };
        'defined';
    """.trimIndent()

    /** The official unary-RPC envelope poster (client-request/server-response). */
    private val postHelper = """
        const post = (endpoint, payload) => fetch('/api/' + endpoint, {
          method: 'POST', credentials: 'same-origin',
          headers: {'content-type': 'application/json'},
          body: JSON.stringify({type: 'client-request',
            rpcId: 'bandroid-probe-' + endpoint + '-' + Math.random().toString(36).slice(2, 8),
            method: endpoint, payload}),
        }).then((r) => r.json());
    """.trimIndent()

    /** Leg 0: wait (bounded) for the official shell to reach its
     * workspace-selection state, then pace ONE quiet window (3s) so the
     * UI's boot-RPC burst is fully observed before the probe's own calls —
     * the manifest's probe-after-UI order is deterministic by construction.
     * The marker matches BOTH upstream locales (the shell's text follows the
     * WebView locale: zh 选择工作区 / en Choose workspace — the iOS sibling
     * only needed the zh row on a zh simulator). The bound is 60s (the iOS
     * sibling's 20s assumed a device-class WebView; the emulator's recovery
     * backoff retries measurably slower — bound, never pace). */
    private val shellSettleLeg = """
        const shellSettle = async () => {
          const deadline = Date.now() + 60000;
          while (Date.now() < deadline) {
            const text = (document.body.innerText || '');
            if (text.includes('$SHELL_MARK') || text.includes('Choose workspace')) {
              await new Promise((r) => setTimeout(r, 3000));
              return 'shell';
            }
            await new Promise((r) => setTimeout(r, 100));
          }
          return 'shell-timeout';
        };
        out.shell = await shellSettle();
    """.trimIndent()

    /** Leg 1: the REAL session.list answer + the reported session identity. */
    private val sessionListLeg = """
        const rpc = await post('session.list', {args: {}});
        out.rpc = rpc && rpc.type === 'server-response'
          ? (rpc.result.ok === true ? 'ok' : 'error:' + (rpc.result.error || {}).code)
          : 'malformed';
        const items = (rpc.result && rpc.result.ok && rpc.result.value && rpc.result.value.items) || [];
        out.sessions = items.length;
        out.blank = items.length > 0 ? !!items[0].blank : null;
        out.running = items.length > 0 ? !!items[0].running : null;
        const sessionId = items.length > 0 ? items[0].sessionId : null;
        out.sessionId = String(sessionId);
    """.trimIndent()

    /** Leg 2: attach the session/journal stream and collect frames until the
     * baseline AND the live turn-2 have both arrived (two assistant/message
     * events, then a quiet window — never timing trust). */
    private val journalLeg = """
        const frames = [];
        out.mux = await new Promise((resolve) => {
          const ws = new WebSocket('ws://' + location.host + '/api/remote.mux');
          const done = (verdict) => { try { ws.close(); } catch (e) {} resolve(verdict); };
          const t = setTimeout(() => done('timeout'), 20000);
          ws.onopen = () => {
            ws.onmessage = (ev) => {
              const frame = JSON.parse(ev.data);
              if (frame.streamId !== '$STREAM_ID') return;
              if (frame.type === 'item') frames.push(frame.value);
              if (frame.type === 'error') { clearTimeout(t); done('error:' + (frame.error || {}).code); }
            };
            ws.send(JSON.stringify({type: 'open', streamId: '$STREAM_ID',
              endpoint: 'session/journal', payload: {args: {address: {sessionId}}}}));
            let lastArrival = Date.now();
            let seen = 0;
            const watch = setInterval(() => {
              const events = frames.filter((f) => f && f.event).map((f) => f.event);
              if (events.length > seen) { lastArrival = Date.now(); seen = events.length; }
              const turnsDone = events.filter((e) => e.type === 'assistant/message').length;
              if (turnsDone >= 2 && Date.now() - lastArrival > 1500) {
                clearInterval(watch); clearTimeout(t); done('frames');
              }
            }, 100);
          };
          ws.onerror = () => { clearTimeout(t); done('ws-error'); };
        });
        const events = frames.filter((f) => f && f.event).map((f) => f.event);
        out.journal = {
          frames: events.length,
          types: events.map((e) => e.type),
          firstSeq: events.length > 0 ? events[0].seq : null,
          lastSeq: events.length > 0 ? events[events.length - 1].seq : null,
        };
    """.trimIndent()

    /** Leg 3: the rendered read — module-system liveness + the TRUE shell
     * state (the upstream DOM, reported as-is). */
    private val renderedLeg = """
        const deadline = Date.now() + 5000;
        while (Date.now() < deadline
            && (!window.__ModuleLoader__ || window.__ModuleLoader__.mode !== 'live')) {
          await new Promise((r) => setTimeout(r, 50));
        }
        out.moduleMode = window.__ModuleLoader__ ? String(window.__ModuleLoader__.mode) : 'none';
        const boot = document.querySelector('[data-dsh-boot]');
        out.page = {boot: boot !== null,
          bootText: boot ? boot.textContent.slice(0, 160) : '',
          rootHasChild: ((document.getElementById('root') || {}).childElementCount || 0) > 0,
          bodyText: (document.body.innerText || '').replace(/\s+/g, ' ').slice(0, 300)};
    """.trimIndent()

    /** Evaluates one script on the page (UI thread required by the WebView). */
    fun evaluate(webView: WebView, script: String) {
        webView.post { webView.evaluateJavascript(script, null) }
    }

    // ---- probe verdict -------------------------------------------------------

    /** The verdict-relevant read of one probe result: either a drive-fatal
     * failure message, or the ordered events to emit plus a pass. */
    sealed class Verdict {
        data class Failure(val message: String) : Verdict()
        data class Pass(val events: List<Pair<String, JSONObject>>) : Verdict()
    }

    fun verdict(json: String): Verdict {
        val probe = try {
            JSONObject(json)
        } catch (_: Exception) {
            return Verdict.Failure("probe returned no parseable result: ${json.take(200)}")
        }
        val page = probe.optJSONObject("page")
            ?: return Verdict.Failure("probe returned no parseable result: ${json.take(200)}")
        failureCause(probe)?.let { return Verdict.Failure(it) }
        val journal = probe.optJSONObject("journal") ?: JSONObject()
        val types = toStringList(journal.optJSONArray("types"))
        return Verdict.Pass(
            listOf(
                "session.list.real" to JSONObject()
                    .put("sessions", probe.optInt("sessions", -1))
                    .put("blank", probe.optBoolean("blank", true))
                    .put("running", probe.optBoolean("running", true))
                    .put("sessionId", probe.optString("sessionId")),
                "session.journal.real" to JSONObject()
                    .put("frames", types.size)
                    .put("baseline", baselineTypes.size)
                    .put("live", liveTypes.size)
                    .put("lastType", types.lastOrNull() ?: ""),
                "module.system.live" to JSONObject()
                    .put("mode", probe.optString("moduleMode"))
                    .put("bootstrap", "@deepseek-ai/dsh-client-modules (vendored)"),
                "page.rendered" to JSONObject()
                    .put("boot", page.optBoolean("boot", true))
                    .put("rootHasChild", page.optBoolean("rootHasChild"))
                    .put("moduleMode", probe.optString("moduleMode"))
                    .put("bodyText", page.optString("bodyText")),
            ),
        )
    }

    /** The ordered hard checks over one probe result; null = all pass. */
    private fun failureCause(probe: JSONObject): String? {
        if (probe.optString("shell") != "shell") {
            return "the official shell never reached its workspace state: " +
                probe.optString("shell", "nil")
        }
        if (probe.optString("rpc") != "ok") {
            return "probe session.list outcome: ${probe.optString("rpc", "nil")}"
        }
        if (probe.optInt("sessions", -1) != 1) {
            return "probe session.list sessions: ${probe.optInt("sessions", -1)}"
        }
        if (probe.optBoolean("blank", true)) {
            return "the reported session is blank (no session log)"
        }
        if (probe.optString("mux") != "frames") {
            return "probe journal outcome: ${probe.optString("mux", "nil")}"
        }
        val journal = probe.optJSONObject("journal") ?: JSONObject()
        val types = toStringList(journal.optJSONArray("types"))
        val expected = baselineTypes + liveTypes
        if (types != expected) {
            return "journal frame types mismatch: ${types.size} frames, " +
                "expected ${expected.size}: $types"
        }
        if (journal.optInt("firstSeq", -1) != 0 ||
            journal.optInt("lastSeq", -1) != expected.size - 1
        ) {
            return "journal seq range not contiguous: " +
                "${journal.optInt("firstSeq", -1)}..${journal.optInt("lastSeq", -1)}"
        }
        val moduleMode = probe.optString("moduleMode", "none")
        if (moduleMode != "live") {
            return "the official module system never went live (mode: $moduleMode)"
        }
        return null
    }

    private fun toStringList(array: JSONArray?): List<String> {
        if (array == null) return emptyList()
        return (0 until array.length()).mapNotNull { runCatching { array.getString(it) }.getOrNull() }
    }
}

File: hosts/ios/App/Source/SessionLiveProbe.swift
import Foundation
import WebKit

/// The b3 same-origin probe (platform-side, M3 rendered-state precedent):
/// the page is untouched upstream code — the probe defines `__b3Run`, posts
/// the REAL session.list RPC through the official envelope, attaches the
/// mux session/journal stream for the session the runtime reports, collects
/// the journal frames (baseline + the live turn-2 stream), and reads the
/// TRUE rendered state. Split from SessionLiveRuntime to keep the drive
/// under the file-size gate.
enum SessionLiveProbe {
    /// Expected journal shape (frozen from the observed on-device runs):
    /// the FIRST scripted turn's full log rides the baseline — header +
    /// context folds included — and the second turn streams LIVE after the
    /// attach with the shorter log (its header/context folds reuse the
    /// turn-1 state). 19 frames, contiguous seqs 0..18.
    static let baselineTypes = [
        "agent/inbox/spliced", "turn/start", "agent/inbox/spliced", "step/start",
        "system/message", "user/message", "request/header", "request/context",
        "assistant/message", "step/end", "turn/end",
    ]
    static let liveTypes = [
        "agent/inbox/spliced", "turn/start", "agent/inbox/spliced", "step/start",
        "user/message", "assistant/message", "step/end", "turn/end",
    ]

    /// Defines `__b3Run` on the page. Step order: real RPC (session.list),
    /// mux upgrade + journal attach for the reported session, frame
    /// collection until two assistant/message events quiet down, then the
    /// rendered read. The legs are separate constants (each stays under the
    /// function-size gate; interpolation assembles the one script).
    static func probeScript() -> String {
        """
        window.__b3Run = async () => {
          const out = {};
          \(postHelper)
          \(shellSettleLeg)
          \(sessionListLeg)
          \(journalLeg)
          \(renderedLeg)
          return JSON.stringify(out);
        };
        'defined';
        """
    }

    /// The official unary-RPC envelope poster (client-request/server-response).
    private static let postHelper = """
        const post = (endpoint, payload) => fetch('/api/' + endpoint, {
          method: 'POST', credentials: 'same-origin',
          headers: {'content-type': 'application/json'},
          body: JSON.stringify({type: 'client-request',
            rpcId: 'b3-probe-' + endpoint + '-' + Math.random().toString(36).slice(2, 8),
            method: endpoint, payload}),
        }).then((r) => r.json());
        """

    /// Leg 0: wait (bounded) for the official shell to reach its
    /// workspace-selection state, then pace ONE quiet window (3s) so the
    /// UI's boot-RPC burst is fully observed before the probe's own calls —
    /// the manifest's probe-after-UI order is deterministic by construction.
    private static let shellSettleLeg = """
        const shellSettle = async () => {
          const deadline = Date.now() + 20000;
          while (Date.now() < deadline) {
            const text = (document.body.innerText || '');
            if (text.includes('\u{9009}\u{62e9}\u{5de5}\u{4f5c}\u{533a}')) {
              await new Promise((r) => setTimeout(r, 3000));
              return 'shell';
            }
            await new Promise((r) => setTimeout(r, 100));
          }
          return 'shell-timeout';
        };
        out.shell = await shellSettle();
        """

    /// Leg 1: the REAL session.list answer + the reported session identity.
    private static let sessionListLeg = """
        const rpc = await post('session.list', {args: {}});
        out.rpc = rpc && rpc.type === 'server-response'
          ? (rpc.result.ok === true ? 'ok' : 'error:' + (rpc.result.error || {}).code)
          : 'malformed';
        const items = (rpc.result && rpc.result.ok && rpc.result.value && rpc.result.value.items) || [];
        out.sessions = items.length;
        out.blank = items.length > 0 ? !!items[0].blank : null;
        out.running = items.length > 0 ? !!items[0].running : null;
        const sessionId = items.length > 0 ? items[0].sessionId : null;
        out.sessionId = String(sessionId);
        """

    /// Leg 2: attach the session/journal stream and collect frames until the
    /// baseline AND the live turn-2 have both arrived (two assistant/message
    /// events, then a quiet window — never timing trust).
    private static let journalLeg = """
        const frames = [];
        out.mux = await new Promise((resolve) => {
          const ws = new WebSocket('ws://' + location.host + '/api/remote.mux');
          const done = (verdict) => { try { ws.close(); } catch (e) {} resolve(verdict); };
          const t = setTimeout(() => done('timeout'), 20000);
          ws.onopen = () => {
            ws.onmessage = (ev) => {
              const frame = JSON.parse(ev.data);
              if (frame.streamId !== 'b3-probe-journal') return;
              if (frame.type === 'item') frames.push(frame.value);
              if (frame.type === 'error') { clearTimeout(t); done('error:' + (frame.error || {}).code); }
            };
            ws.send(JSON.stringify({type: 'open', streamId: 'b3-probe-journal',
              endpoint: 'session/journal', payload: {args: {address: {sessionId}}}}));
            let lastArrival = Date.now();
            let seen = 0;
            const watch = setInterval(() => {
              const events = frames.filter((f) => f && f.event).map((f) => f.event);
              if (events.length > seen) { lastArrival = Date.now(); seen = events.length; }
              const turnsDone = events.filter((e) => e.type === 'assistant/message').length;
              if (turnsDone >= 2 && Date.now() - lastArrival > 1500) {
                clearInterval(watch); clearTimeout(t); done('frames');
              }
            }, 100);
          };
          ws.onerror = () => { clearTimeout(t); done('ws-error'); };
        });
        const events = frames.filter((f) => f && f.event).map((f) => f.event);
        out.journal = {
          frames: events.length,
          types: events.map((e) => e.type),
          firstSeq: events.length > 0 ? events[0].seq : null,
          lastSeq: events.length > 0 ? events[events.length - 1].seq : null,
        };
        """

    /// Leg 3: the rendered read — module-system liveness + the TRUE shell
    /// state (the upstream DOM, reported as-is).
    private static let renderedLeg = """
        const deadline = Date.now() + 5000;
        while (Date.now() < deadline
            && (!window.__ModuleLoader__ || window.__ModuleLoader__.mode !== 'live')) {
          await new Promise((r) => setTimeout(r, 50));
        }
        out.moduleMode = window.__ModuleLoader__ ? String(window.__ModuleLoader__.mode) : 'none';
        const boot = document.querySelector('[data-dsh-boot]');
        out.page = {boot: boot !== null,
          bootText: boot ? boot.textContent.slice(0, 160) : '',
          rootHasChild: ((document.getElementById('root') || {}).childElementCount || 0) > 0,
          bodyText: (document.body.innerText || '').replace(/\\s+/g, ' ').slice(0, 300)};
        """

    /// Evaluates one script on the page (main thread).
    static func evaluate(
        _ webView: WKWebView, _ script: String,
        completion: @escaping (Any?, Error?) -> Void
    ) {
        DispatchQueue.main.async {
            webView.evaluateJavaScript(script) { result, error in
                completion(result, error)
            }
        }
    }

    /// Runs one async page expression and AWAITS its promise (plain
    /// evaluateJavaScript would return the Promise object itself, which the
    /// bridge refuses — WKError 5 "unsupported type").
    static func awaitPromise(
        _ webView: WKWebView, _ expression: String,
        completion: @escaping (Any?, Error?) -> Void
    ) {
        DispatchQueue.main.async {
            webView.callAsyncJavaScript(
                "return await (\(expression));",
                arguments: [:], in: nil, in: .page
            ) { result in
                switch result {
                case .success(let value): completion(value, nil)
                case .failure(let error): completion(nil, error)
                }
            }
        }
    }

    // ---- probe verdict -------------------------------------------------------

    /// The verdict-relevant read of one probe result: either a drive-fatal
    /// failure message, or the ordered events to emit plus a pass.
    enum Verdict {
        case failure(String)
        case pass(events: [(String, [String: Any])])
    }

    static func verdict(_ json: String) -> Verdict {
        guard let data = json.data(using: .utf8),
              let probe = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any],
              let page = probe["page"] as? [String: Any] else {
            return .failure("probe returned no parseable result: \(json.prefix(200))")
        }
        guard checkSessionList(probe) else {
            return .failure(journalError(probe))
        }
        guard let journal = checkJournal(probe) else {
            return .failure(journalError(probe))
        }
        // The official module system went live (REAL boot progression).
        let moduleMode = probe["moduleMode"] as? String ?? "none"
        guard moduleMode == "live" else {
            return .failure("the official module system never went live (mode: \(moduleMode))")
        }
        let types = journal["types"] as? [String] ?? []
        return .pass(events: [
            ("session.list.real", [
                "sessions": probe["sessions"] ?? -1,
                "blank": probe["blank"] ?? true,
                "running": probe["running"] ?? true,
                "sessionId": probe["sessionId"] ?? "",
            ]),
            ("session.journal.real", [
                "frames": types.count,
                "baseline": baselineTypes.count,
                "live": liveTypes.count,
                "lastType": types.last ?? "",
            ]),
            ("module.system.live", ["mode": moduleMode,
                "bootstrap": "@deepseek-ai/dsh-client-modules (vendored)"]),
            ("page.rendered", [
                "boot": (page["boot"] as? Bool) ?? false,
                "rootHasChild": (page["rootHasChild"] as? Bool) ?? false,
                "moduleMode": moduleMode,
                "bodyText": (page["bodyText"] as? String) ?? "",
            ]),
        ])
    }

    /// Leg 1 checks: the REAL session.list answer — ok, exactly the
    /// configured agent session, non-blank (the turn's log exists).
    private static func checkSessionList(_ probe: [String: Any]) -> Bool {
        probe["shell"] as? String == "shell"
            && probe["rpc"] as? String == "ok"
            && probe["sessions"] as? Int == 1
            && (probe["blank"] as? Bool) == false
    }

    /// Leg 2 checks: the REAL journal — the first turn's full log in the
    /// baseline, then turn 2 streamed live, contiguous seqs. Returns nil
    /// when a check failed (the reason is in journalError).
    private static func checkJournal(_ probe: [String: Any]) -> [String: Any]? {
        guard probe["mux"] as? String == "frames",
              let journal = probe["journal"] as? [String: Any],
              let types = journal["types"] as? [String],
              types == baselineTypes + liveTypes,
              journal["firstSeq"] as? Int == 0,
              journal["lastSeq"] as? Int == baselineTypes.count + liveTypes.count - 1
        else { return nil }
        return journal
    }

    /// The first failed leg's diagnosis, in manifest order.
    private static func journalError(_ probe: [String: Any]) -> String {
        if probe["shell"] as? String != "shell" {
            return "the official shell never reached its workspace state: "
                + "\(probe["shell"] ?? "nil")"
        }
        if probe["rpc"] as? String != "ok" {
            return "probe session.list outcome: \(probe["rpc"] ?? "nil")"
        }
        if probe["sessions"] as? Int != 1 {
            return "probe session.list sessions: \(probe["sessions"] ?? -1)"
        }
        if (probe["blank"] as? Bool) != false {
            return "the reported session is blank (no session log)"
        }
        if probe["mux"] as? String != "frames" {
            return "probe journal outcome: \(probe["mux"] ?? "nil")"
        }
        if let journal = probe["journal"] as? [String: Any],
           let types = journal["types"] as? [String] {
            if types != baselineTypes + liveTypes {
                return "journal frame types mismatch: \(types.count) frames, "
                    + "expected \((baselineTypes + liveTypes).count): \(types)"
            }
            return "journal seq range not contiguous: "
                + "\(journal["firstSeq"] ?? -1)..\(journal["lastSeq"] ?? -1)"
        }
        return "probe journal frames missing"
    }
}
